File: test/Test0.js
'use strict'
const xml = require('../xml-csharp-cereal');
const Long = require("long"); // to test simple decode/encode overrides

// Just a big mash up of stuff for XmlSerializer

module.exports.GetFactory = function()
{
    return (new xml.XmlTemplateFactory(TestClass, SubTestClass))
        //.addEnum('MyEnum', MyEnumSimple)
        .addEnum('MyEnum', MyEnumExplicit)
        //.addDict('SerializableDictionaryOfStringInt32','KeyValuePair', new xml.XmlTemplateItem('Key','string'),new xml.XmlTemplateItem('Value','int'))
        .addDict('SerializableDictionaryOfStringInt32','KeyValuePair',['Key','string'],['Value','int'])
        .addDict('SerializableDictionaryTypedOfStringInt32','KeyValuePair',['Key','string'],['Value','int'], null, true)
        // is same as .addDictQuick('SerializableDictionaryOfStringInt32','int')
        .addDictQuick('SerializableDictionaryOfStringSerializableDictionaryOfStringInt32','SerializableDictionaryOfStringInt32')
        .addDictQuick('SerializableDictionaryOfStringSubTestClass','SubTestClass')
        .addDictQuick('SerializableDictionaryOfStringArrayOfMyEnum',['Value','MyEnum',1])
        .addDictQuick('SerializableDictionaryOfStringArrayOfSerializableDictionaryOfStringInt32',['Value','SerializableDictionaryOfStringInt32', 1])
        // defaults for long are to decode/encode as string, override with Long.js for this test
        .setSimpleCodec(['long', 'Int64'], decodeLongjs, encodeLongjs)
        .setSimpleCodec(['ulong', 'UInt64'], decodeULongjs, encodeULongjs);
}

class TestClass
{
    constructor()
    {
        // This constructor should mirror its C# counterpart, otherwise missing tags might result in differences due to mismatched defaults.
        // If the C# constructor leaves something null, also null it here.
        // If the C# constructor inits a member, do so here.
        this.SomeAttr = null; // [XmlAttribute] public string SomeAttr;
        this.MyEnumProp = 0; //public MyEnum MyEnumProp; // C# inits enums as zero
        this.MyEnumArray=null; //public MyEnum[] MyEnumArray;
        this.MyBool=false;
        this.MyByte=0; //public byte MyByte;
        this.MyInt=0; //public int MyInt;
        this.MyUInt=0; //public uint MyUInt;
        this.MyLong=0; //public long MyLong;
        this.MyDouble=0; //public double MyDouble;
        this.MyNullInt; //public int? MyNullInt;
        // C# Bug?: if int?[] occurs before int[][], int[][] becomes "ArrayOfInt1" instead of "ArrayOfInt" ?
        // This bug can be avoided by using [XmlArrayItem(ElementName = "ArrayOfInt", IsNullable = false, Type = typeof(int[]))] on MyJagIntArray
        this.MyNullIntArr; //public int?[] MyNullIntArr;
        this.MyIntArray=null; //public int[] MyIntArray;
        this.MyJagIntArray=null; //public int[][] MyJagIntArray;
        this.MyJagIntArray2=null; //public int[][][] MyJagIntArray2;
        this.MyJagEnumArray=null; //public MyEnum[][] MyJagEnumArray;
        this.MyStrList=null; //public List<string> MyStrList; - NOTE: it seems XmlSerializer will create a list on the other side regardless of whether a tag is actually present in XML

        this.MySubClass=new SubTestClass(); //public SubTestClass MySubClass
        this.MySubClassArray=null; //public SubTestClass[] MySubClassArray;
        this.MyIntDict=[]; //public SerializableDictionary<string, int> MyIntDict
        this.MyIntDictTyped=[]; //public SerializableDictionaryTyped<string, int> MyIntDictTyped
        this.MyIntDictArr=null; //public SerializableDictionary<string, int>[] MyIntDictArr;
        this.MySubClassDict=[]; //public SerializableDictionary<string, SubTestClass> MySubClassDict
        this.MyEnumArrDict = null; //public SerializableDictionary<string, MyEnum[]> MyEnumArrDict;
        this.MyIntDictDict = null; //public SerializableDictionary<string, SerializableDictionary<string, int> > MyIntDictDict;
        this.MyIntDictArrDict = null; //public SerializableDictionary<string, SerializableDictionary<string, int>[]> MyIntDictArrDict;

        this.MyTime = null; // DateTime
        this.MySpan = null; // TimeSpan
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        temp.addString('SomeAttr').attr();
        temp.add('MyEnumProp', 'MyEnum');
        temp.add('MyEnumArray', 'MyEnum', 1);
        temp.addBool('MyBool');
        temp.addByte('MyByte');
        temp.addInt('MyInt');
        temp.addUInt('MyUInt');
        temp.addLong('MyLong');
        temp.addDouble('MyDouble');
        temp.addInt('MyJagIntArray2', 3);
        temp.addInt('MyNullInt').nullable();
        temp.addInt('MyNullIntArr', 1).nullable();
        temp.addInt('MyIntArray', 1)
        //temp.addInt('MyJagIntArray', 2); // jagged array has two dimensions
        temp.addInt('MyJagIntArray', ['int','ArrayOfInt']);
        temp.add('MyJagEnumArray', 'MyEnum', 2);
        temp.addString('MyStrList', 1);
        temp.add('MySubClass', 'SubTestClass');
        temp.add('MySubClassArray', 'SubTestClass', 1);
        temp.add('MyIntDictTyped', 'SerializableDictionaryTypedOfStringInt32');
        temp.add('MyIntDict', 'SerializableDictionaryOfStringInt32');
        temp.add('MyIntDictArr', 'SerializableDictionaryOfStringInt32', 1);
        temp.add('MySubClassDict', 'SerializableDictionaryOfStringSubTestClass');
        // complex embedding?
        temp.add('MyEnumArrDict', 'SerializableDictionaryOfStringArrayOfMyEnum');
        temp.add('MyIntDictDict', 'SerializableDictionaryOfStringSerializableDictionaryOfStringInt32');
        temp.add('MyIntDictArrDict', 'SerializableDictionaryOfStringArrayOfSerializableDictionaryOfStringInt32');
        temp.addDateTime('MyTime');
        temp.addTimeSpan('MySpan');
        return temp;
    }
}
module.exports.TestClass=TestClass;

class SubTestClass
{
    constructor()
    {
        this.SubAttr=0; // [XmlAttribute] public int SubAttr;
        this.SubAttr2=null; // [XmlAttribute] public string SubAttr2;
        this.MySubInt=0; //public int MySubInt;
        this.MySubStr=null; //public string MySubStr;
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        temp.addInt('SubAttr').attr();
        temp.addString('SubAttr2').attr();
        temp.addInt('MySubInt');
        temp.addString('MySubStr');
        return temp;
    }
}
module.exports.SubTestClass=SubTestClass;

// Two ways to provide an enum definition
const MyEnumSimple = { zero:0, one:1, two:2, three:3 }
const MyEnumExplicit =
{
    getEnumValue: function(n)
    {
        var lut = { zero:0, one:1, two:2, three:3 };
        if (lut[n]==undefined) return null; // decode will likely throw if raw value cannot be determined
        return lut[n];
    },
    getEnumName: function(v)
    {
        var lut = { 0:'zero', 1:'one', 2:'two', 3:'three' }
        if (lut[v]==undefined) return v;  // could return raw value, or return null and encode will throw Error
        return lut[v];
    }
}

// examples of decode/encode overrides (using Long.js for 64 bit integers)
function decodeLongjs(val)
{
    // parse XML value into Long instance
    if (val == null) return null; // handle nullable
    return Long.fromValue(val, false); // let it throw on error
}
function encodeLongjs(val)
{
    // output Long as string for XML node
    if (val == null) return null; // handle nullable
    // if not Long instance already, try to parse into a Long
    val = Long.fromValue(val, false);
    return val.toString(); // let it throw on error
}
function decodeULongjs(val)
{
    // parse XML value into Long instance
    if (val == null) return null; // handle nullable
    return Long.fromValue(val, true); // let it throw on error
}
function encodeULongjs(val)
{
    // output Long as string for XML node
    if (val == null) return null; // handle nullable
    // if not Long instance already, try to parse into a Long
    val = Long.fromValue(val, true);
    return val.toString(); // let it throw on error
}

File: test/csharpxml/Test0.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml.Serialization;

/*
 * This a test class to serialize.
 * It is the default test case, or test number zero.
 * 
 */

namespace csharpxml.Test0
{
    // NOTE: if enum does not handle zero and enum prop is not given initial non-zero value, serializer throws.
    public enum MyEnum { zero=0, one=1, two=2, three=3 };
    public class TestClass : XmlTestObject<TestClass>
    {
        [XmlAttribute]
        public string SomeAttr;
        public MyEnum MyEnumProp;
        public MyEnum[] MyEnumArray;
        public bool MyBool;
        public byte MyByte;
        public int MyInt;
        public uint MyUInt;
        public long MyLong;
        public double MyDouble;
        public int[][][] MyJagIntArray2; // moving this up here to avoid tag renaming by nullables below
        public int? MyNullInt;
        public int?[] MyNullIntArr; // NOTE: if this nullable array occurs before MyJagIntArray, MyJagIntArray's tags become "ArrayOfInt1" instead of "ArrayOfInt"
        public int[] MyIntArray;
        [XmlArrayItem(ElementName = "ArrayOfInt", IsNullable = false, Type = typeof(int[]))] // this prevents bug described above
        public int[][] MyJagIntArray;
        public MyEnum[][] MyJagEnumArray;
        public List<string> MyStrList = new List<string>(); // fun fact, XmlSerializer will construct one regardless of whether any tag is there

        public SubTestClass MySubClass = new SubTestClass();
        public SubTestClass[] MySubClassArray;
        public SerializableDictionaryTyped<string, int> MyIntDictTyped = new SerializableDictionaryTyped<string, int>();
        public SerializableDictionary<string, int> MyIntDict = new SerializableDictionary<string, int>();
        public SerializableDictionary<string, int>[] MyIntDictArr;
        public SerializableDictionary<string, SubTestClass> MySubClassDict = new SerializableDictionary<string, SubTestClass>();
        public SerializableDictionary<string, MyEnum[]> MyEnumArrDict;
        public SerializableDictionary<string, SerializableDictionary<string, int> > MyIntDictDict;
        public SerializableDictionary<string, SerializableDictionary<string, int>[]> MyIntDictArrDict;

        public SerializableDateTime MyTime;
        public SerializableTimeSpan MySpan;
        /*[XmlArrayItem(ElementName = "DateTime", Type = typeof(SerializableDateTime))]
        public SerializableDateTime[] MyTimes;*/

        public override void PopulateTestObject(int subtest)
        {
            switch (subtest)
            {
                case 0:
                    // run with constructor defaults
                    break;
                case 1:
                    //var tm = new DateTime(2018, 12, 25, 13, 30, 15, 750, DateTimeKind.Utc);
                    //var tm = new DateTime(2018, 12, 25, 13, 30, 15, 750, DateTimeKind.Local);
                    var tm = new DateTime(2018, 12, 25, 13, 30, 15, 750);
                    MyTime = new SerializableDateTime(tm);
                    MySpan = new SerializableTimeSpan(TimeSpan.FromMilliseconds(123456789));
                    /*MyTimes = new SerializableDateTime[]
                    {
                        new SerializableDateTime("5/30/2018 5:30PM"),
                        new SerializableDateTime("5/30/2018 6:30PM")
                    };*/
                    SomeAttr = "att";
                    MyEnumProp = MyEnum.three;
                    MyEnumArray = new MyEnum[] { MyEnum.one, MyEnum.two };
                    MyBool = true;
                    MyInt = 1;
                    MyNullInt = 42;
                    MyNullIntArr = new int?[] { 43, null, 45 };
                    MyByte = byte.MaxValue;
                    MyUInt = uint.MaxValue;
                    MyLong = long.MaxValue;
                    MyDouble = double.MaxValue;
                    MyIntArray = new int[] { 3, 4, 5 };
                    MyJagIntArray = new int[][]
                    {
                        new int[] { 6, 7, 8 },
                        new int[] { 9, 10 },
                    };
                    MyJagIntArray2 = new int[][][]
                    {
                        new int[][]{ new int[] {11,22}, new int[] {33,44} },
                        new int[][]{ new int[] {55,66}, new int[] {77,88} },
                    };
                    MyJagEnumArray = new MyEnum[][]
                    {
                        new MyEnum[] { MyEnum.one, MyEnum.two},
                        new MyEnum[] { MyEnum.three }
                    };
                    MyStrList = new List<string>();
                    MyStrList.Add("List0");
                    MyStrList.Add("List1");
                    MySubClass.Populate(-1);
                    MySubClassArray = new SubTestClass[3];
                    for (int i = 0; i < MySubClassArray.Length-1; ++i)
                    {
                        MySubClassArray[i] = new SubTestClass();
                        MySubClassArray[i].Populate(i);
                    }
                    MyIntDict.Add("first", 55);
                    MyIntDict.Add("second", 65);
                    MyIntDictTyped.Add("third", 75);
                    MyIntDictTyped.Add("fourth", 85);
                    MyIntDictArr = new SerializableDictionary<string, int>[2];
                    for (int i = 0; i < MyIntDictArr.Length; ++i)
                    {
                        MyIntDictArr[i] = new SerializableDictionary<string, int>();
                        MyIntDictArr[i].Add("item1-" + i.ToString(), i * 9);
                        MyIntDictArr[i].Add("item2-" + i.ToString(), i * 19);
                    }
                    var sub = new SubTestClass();
                    sub.Populate(11);
                    MySubClassDict.Add("subclass", sub);
                    MyEnumArrDict = new SerializableDictionary<string, MyEnum[]>();
                    MyEnumArrDict.Add("EnumKey", new MyEnum[] { MyEnum.three, MyEnum.two, MyEnum.one });
                    // good lord
                    MyIntDictDict = new SerializableDictionary<string, SerializableDictionary<string, int>>();
                    MyIntDictDict.Add("really", MyIntDict);
                    // stop, please
                    MyIntDictArrDict = new SerializableDictionary<string, SerializableDictionary<string, int>[]>();
                    MyIntDictArrDict.Add("yes", MyIntDictArr);
                    break;
                default:
                    break;
            }
        }
    } // END CLASS: TestClass

    public class SubTestClass
    {
        [XmlAttribute]
        public int SubAttr;
        [XmlAttribute]
        public string SubAttr2;
        /* serializer doesn't do nullable attributes
        [XmlAttribute]
        public int? SubAttr3;*/
        public int MySubInt;
        public string MySubStr;

        public void Populate(int test)
        {
            SubAttr = test * 9;
            SubAttr2 = "att_" + test.ToString();
            //SubAttr3 = test * 42;
            MySubInt = test * 30 + 15;
            MySubStr = "SubClass string " + test.ToString();
        }
    } // END CLASS: SubTestClass

} // END NAMESPACE
